ci: Migrate from stateful to stateless runner and apply optimizations to reduce workflow runtime:

jackluo923 opened the pull request:
<!-- markdownlint-disable MD012 -->

<!--
Set the PR title to a meaningful commit message that:

* is in imperative form.
* follows the Conventional Commits specification (https://www.conventionalcommits.org).
  * See https://github.com/commitizen/conventional-commit-types/blob/master/index.json for possible
    types.

Example:

fix: Don't add implicit wildcards ('*') at the beginning and the end of a query (fixes #390).
-->

# Description

<!-- Describe what this request will change/fix and provide any details necessary for reviewers. -->

Replaced stateful self-hosted runner builds with ephemeral containerized execution using pre-built Docker images with cached dependencies, eliminating state accumulation and dependency conflicts while improving build reliability and test execution performance.

Migration from ubuntu-release (stateful direct runner) to containerized workflow:

1. Introduced Docker builder image for ephemeral execution:
   - Created yscope-velox-builder-img.yml to build and publish builder image
   - Image includes all dependencies and pre-populated ccache from warmup build
   - Cache key based on hash of scripts/** - image only rebuilt when dependencies change
   - Most builds reuse existing cached image, avoiding ~30min image build overhead
   - Published to ghcr.io with dynamic tagging for parallel workflow execution
   - Each job runs in fresh container with consistent, known state

2. Containerized build and test execution:
   - Replaced stateful runner with ephemeral container-based jobs
   - Aligned paths to GitHub Actions workspace (/__w/velox/velox)
   - Configured ccache with BASEDIR matching workspace for proper cache hits
   - Disabled ccache compression (COMPRESSLEVEL=0) for faster CI execution
   - Increased parallelism for self-hosted runners (16 high-mem jobs, 12 link jobs)
   - Containers automatically removed after job completion

3. Split tests into two parallel jobs (instead of single sequential job):
   - velox-exec-test: Runs only velox_exec_test (~16 minutes, the bottleneck)
   - other-tests: Runs all remaining tests in parallel
   - Both jobs build independently in separate ephemeral containers
   - Shared ccache provides fast builds despite independent execution
   - Total time ~25 minutes with better failure visibility

4. Test execution configuration:
   - Added ulimit -n 8192 for velox_table_evolution_fuzzer_test
   - Documented pattern for separating additional slow tests in future

5. CI configuration improvements:
   - Added workflow_dispatch trigger for manual workflow runs
   - Configured zizmor suppressions for intentionally unpinned images
   - Improved CMake installation documentation in Dockerfile

Benefits of ephemeral containerized approach vs stateful runner:
- Consistent, reproducible build environment (no state accumulation)
- Eliminates dependency conflicts from previous runs
- Faster builds via pre-populated ccache in builder image
- Image reused across most builds - only rebuilt when dependencies change
- Better resource utilization (parallel test execution in separate containers)
- Improved failure visibility (separate slow vs fast test jobs)
- No need to install dependencies on runner for each run
- Automatic cleanup (containers destroyed after job completion)
- Dynamic image caching reduces unnecessary rebuilds
- Scalable pattern for adding more parallel test jobs

This migration modernizes the CI pipeline from stateful runner execution to ephemeral containerized builds with optimized test parallelization.

# Checklist

<!-- Ensure each item below is satisfied and indicate so by inserting an `x` within each `[ ]`. -->

* [x] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [x] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [x] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

<!-- Describe what tests and validation you performed on the change. -->

Ran the CI workflows multiple times with success. For example:
- https://github.com/y-scope/velox/actions/runs/19681868579/job/56381881616


[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html





<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Reusable builder image and parallel build-and-test matrix with two test groups, per-group test args and ulimit, plus optional Clang selection.

* **Chores**
  * Split CI into image-build and build-and-test stages, use builder-image outputs to speed runs, adaptive MAKEFLAGS for runner type, cache alignment, updated concurrency and package-publish permission, and skip-if-exists image publishing.

* **Documentation**
  * Added ignore rule and explanatory comments for image publishing and workflow behaviour.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jackluo923 committed:
ci: migrate from stateful runner execution to ephemeral containerized builds

Replaced stateful self-hosted runner builds with ephemeral containerized
execution using pre-built Docker images with cached dependencies, eliminating
state accumulation and dependency conflicts while improving build reliability
and test execution performance.

Migration from ubuntu-release (stateful direct runner) to containerized workflow:

1. Introduced Docker builder image for ephemeral execution:
   - Created yscope-velox-builder-img.yml to build and publish builder image
   - Image includes all dependencies and pre-populated ccache from warmup build
   - Cache key based on hash of scripts/** - image only rebuilt when dependencies change
   - Most builds reuse existing cached image, avoiding ~30min image build overhead
   - Published to ghcr.io with dynamic tagging for parallel workflow execution
   - Each job runs in fresh container with consistent, known state

2. Containerized build and test execution:
   - Replaced stateful runner with ephemeral container-based jobs
   - Aligned paths to GitHub Actions workspace (/__w/velox/velox)
   - Configured ccache with BASEDIR matching workspace for proper cache hits
   - Disabled ccache compression (COMPRESSLEVEL=0) for faster CI execution
   - Increased parallelism for self-hosted runners (16 high-mem jobs, 12 link jobs)
   - Containers automatically removed after job completion

3. Split tests into two parallel jobs (instead of single sequential job):
   - velox-exec-test: Runs only velox_exec_test (~16 minutes, the bottleneck)
   - other-tests: Runs all remaining tests in parallel
   - Both jobs build independently in separate ephemeral containers
   - Shared ccache provides fast builds despite independent execution
   - Total time ~25 minutes with better failure visibility

4. Test execution configuration:
   - Added ulimit -n 8192 for velox_table_evolution_fuzzer_test
   - Documented pattern for separating additional slow tests in future

5. CI configuration improvements:
   - Added workflow_dispatch trigger for manual workflow runs
   - Configured zizmor suppressions for intentionally unpinned images
   - Improved CMake installation documentation in Dockerfile

Benefits of ephemeral containerized approach vs stateful runner:
- Consistent, reproducible build environment (no state accumulation)
- Eliminates dependency conflicts from previous runs
- Faster builds via pre-populated ccache in builder image
- Image reused across most builds - only rebuilt when dependencies change
- Better resource utilization (parallel test execution in separate containers)
- Improved failure visibility (separate slow vs fast test jobs)
- No need to install dependencies on runner for each run
- Automatic cleanup (containers destroyed after job completion)
- Dynamic image caching reduces unnecessary rebuilds
- Scalable pattern for adding more parallel test jobs

This migration modernizes the CI pipeline from stateful runner execution to
ephemeral containerized builds with optimized test parallelization.

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,81 @@
+# CI workflows will check out fresh source code at runtime into this same directory.
+RUN rm -rf /__w/velox/velox/*
+
+WORKDIR /__w/velox/velox
```
good point.

jackluo923 replied:
Done - the redundant WORKDIR was removed in commit da16d12ec.

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,81 @@
+# Remove source files but keep ccache
+# The compiled binaries and source are deleted - we only need the ccache artifacts.
+# CI workflows will check out fresh source code at runtime into this same directory.
+RUN rm -rf /__w/velox/velox/*
```
Confident in the path stability, modified as suggested.

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,81 @@
+# CI workflows will check out fresh source code at runtime into this same directory.
+RUN rm -rf /__w/velox/velox/*
+
+WORKDIR /__w/velox/velox
```
unnecessary.

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,81 @@
+RUN wget https://github.com/Kitware/CMake/releases/download/v3.28.3/cmake-3.28.3-linux-x86_64.sh \
+ && chmod +x cmake-3.28.3-linux-x86_64.sh \
+ && ./cmake-3.28.3-linux-x86_64.sh --skip-license --prefix=/usr/local \
+ && rm cmake-3.28.3-linux-x86_64.sh
```
added progress

jackluo923 committed:
ci: minor dockerfile and zizmor config improvements

- Remove redundant WORKDIR directive in builder dockerfile
- Add progress indicator for wget during CMake download
- Add confirmation echo after source cleanup
- Document zizmor rule ignore rationale for unpinned images/actions

jackluo923 committed:
Merge branch 'presto-0.293-clp-connector' into ci

jackluo923 committed:
ci: improve ccache portability and dockerfile layer caching

- Set CCACHE_BASEDIR dynamically at runtime via $GITHUB_WORKSPACE instead
  of hardcoding /__w/velox/velox path, making ccache portable across
  different checkout locations
- Copy only scripts directory first for dependency installation to
  optimize Docker layer caching - source changes no longer invalidate
  the dependency layer
- Move venv to /opt/velox-venv and ccache to /var/cache/ccache for
  cleaner filesystem organization
- Use /tmp for warmup build source to clarify it's temporary
- Clean up temp files after dependency installation and warmup build

jackluo923 committed:
ci: refactor parallel test jobs into matrix strategy

- Consolidate duplicate velox-exec-test and other-tests jobs into single
  build-and-test job with matrix strategy
- Move MAKEFLAGS from step-level to job-level env to reduce duplication
- Parameterize ctest filter args via matrix (ctest-args)
- Add optional ulimit-nofile matrix parameter for tests needing higher
  file descriptor limits

jackluo923 committed:
ci: move ccache stats clearing to dockerfile

Clear ccache statistics after warmup build in dockerfile so the image
starts with zeroed stats. Removes redundant step from workflow.

jackluo923 committed:
ci: fix concurrency group to cancel duplicate push/PR runs

Use github.head_ref for pull_request events so push and PR triggers
for the same branch share the same concurrency group, cancelling
redundant runs.

jackluo923 committed:
ci: fix shell syntax to be POSIX-compatible

Replace bash-specific [[ ]] with POSIX [ ] for compatibility with
GitHub Actions' default /bin/sh shell.

jackluo923 committed:
ci: use runner.environment instead of runner.labels for self-hosted detection

runner.labels is not a valid GitHub Actions context property.
runner.environment is documented and equals 'self-hosted' for self-hosted runners.

jackluo923 commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,72 @@ on:
+          # MAX_LINK_JOBS: Parallel linking tasks (memory-intensive)
+          MAKEFLAGS: >-
+            MAX_HIGH_MEM_JOBS=${{ contains(runner.labels, 'self-hosted') && '16' || '4' }}
+            MAX_LINK_JOBS=${{ contains(runner.labels, 'self-hosted') && '12' || '3' }}
```
Concern addressed.

jackluo923 committed:
ci: fix concurrency group to properly cancel duplicate push/PR runs

Use github.head_ref || github.ref_name pattern so both push and
pull_request events for the same branch resolve to the same group.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+
+# Copy only scripts first for dependency installation.
+# This optimizes Docker layer caching: changes to other source files won't invalidate
+# the dependency installation layer - it only rebuilds when scripts change.
```
```suggestion
# First, copy the dependency installation scripts to optimize Docker layer caching
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# NOTE: setup-ubuntu.sh installs cmake via pip, but pip-installed cmake is not reliably found
+# in container environments, causing FastFloat library build failures. Using the official
+# binary installer ensures cmake is available and dependencies build correctly in the current
+# CI containerized environment.
```
```suggestion
# Install CMake 3.28.3 using its install script
# NOTE: `scripts/setup-ubuntu.sh` installs CMake via pip, but sometimes the pip-installed CMake
# doesn't show up on the path in container environments. Using CMake's install script avoids this
# issue.
```

jackluo923 replied:
How about 
```
# Install CMake 3.28.3 using its install script
# NOTE: setup-ubuntu.sh installs cmake via pip, but sometimes the pip-installed CMake
# doesn't show up on the path in container environments, causing FastFloat library build
# failures. Using CMake's install script avoids this issue.
```
this would retain the blurb about FastFloat. If someone wants to fix the issue later, they don't need experience as much pain of isolating which part of the setup-ubuntu.sh to debug.

kirkrodrigues replied:
```suggestion
# Install CMake 3.28.3 using its install script
# NOTE: `scripts/setup-ubuntu.sh` installs CMake via pip, but sometimes the pip-installed CMake
# doesn't show up on the path in container environments (causing, for example, FastFloat library
# build 1failures). Using CMake's install script avoids this issue.
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# in container environments, causing FastFloat library build failures. Using the official
+# binary installer ensures cmake is available and dependencies build correctly in the current
+# CI containerized environment.
+RUN wget --progress=dot:giga https://github.com/Kitware/CMake/releases/download/v3.28.3/cmake-3.28.3-linux-x86_64.sh \
```
Why not use `curl`?

jackluo923 replied:
It's a personal preference and also a practical one. In many cases, wget is pre-installed while curl is not. Is there a strong reason to use curl? The author of curl suggested to use whichever one that gets the job done: https://daniel.haxx.se/docs/curl-vs-wget.html

kirkrodrigues replied:
* `curl` is installed in `ghcr.io/myoung34/docker-github-actions-runner:ubuntu-jammy`, so the installation shouldn't be a concern.
* We use `curl` in our other scripts at YScope, so the syntax is more well known.

jackluo923 replied:
Switched to curl.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+RUN wget --progress=dot:giga https://github.com/Kitware/CMake/releases/download/v3.28.3/cmake-3.28.3-linux-x86_64.sh \
+ && chmod +x cmake-3.28.3-linux-x86_64.sh \
+ && ./cmake-3.28.3-linux-x86_64.sh --skip-license --prefix=/usr/local \
+ && rm cmake-3.28.3-linux-x86_64.sh
```
We should indent by 4, not 2:

```suggestion
   && chmod +x cmake-3.28.3-linux-x86_64.sh \
   && ./cmake-3.28.3-linux-x86_64.sh --skip-license --prefix=/usr/local \
   && rm cmake-3.28.3-linux-x86_64.sh
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+ && rm cmake-3.28.3-linux-x86_64.sh
+
+# Run the setup script to install all dependencies, then move venv to /opt and clean up
+# Note: setup-ubuntu.sh creates venv at SCRIPTDIR/../.venv (i.e., /tmp/.venv)
```
The code indicates the same thing so I think this comment is a bit redundant.

```suggestion
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# Note: setup-ubuntu.sh creates venv at SCRIPTDIR/../.venv (i.e., /tmp/.venv)
+RUN /tmp/velox-deps/setup-ubuntu.sh \
+ && mv /tmp/.venv /opt/velox-venv \
+ && rm -rf /tmp/velox-deps
```
```suggestion
   && mv /tmp/.venv /opt/velox-venv \
   && rm -rf /tmp/velox-deps
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+ && mv /tmp/.venv /opt/velox-venv \
+ && rm -rf /tmp/velox-deps
+
+# Set up environment to use the Python venv
```
```suggestion
# Activate the virtual environment
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+
+# Set up environment to use the Python venv
+ENV PATH="/opt/velox-venv/bin:${PATH}"
+ENV VIRTUAL_ENV="/opt/velox-venv"
```
Would be helpful to have a comment explaining why we can't do this by sourcing `/opt/velox-venv/bin/activate`.

jackluo923 replied:
done

jackluo923 replied:
The current comment already explains why we can't simply source `/opt/velox-venv/bin/activate` - it's because each RUN command starts a fresh shell, so activation wouldn't persist. Using ENV variables bakes the venv path into the image. I think this explanation is valuable for future maintainers, so I'd prefer to keep the full comment rather than shortening it.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# CCACHE_MAX_SIZE: 5GB quota (actual usage typically a few hundred MB).
+ENV CCACHE_DIR=/var/cache/ccache
+ENV CCACHE_COMPRESSLEVEL=0
+ENV CCACHE_MAX_SIZE=5G
```
```suggestion
ENV CCACHE_DIR=/var/cache/ccache

# Disable compression to trade disk space for speed in CI builds
ENV CCACHE_COMPRESSLEVEL=0
ENV CCACHE_MAX_SIZE=5G
```

jackluo923 replied:
pending on the resolve of other concerns below

jackluo923 replied:
Applied with modification: used `CCACHE_MAXSIZE` (the correct env var name) instead of `CCACHE_MAX_SIZE`.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+ENV CCACHE_DIR=/var/cache/ccache
+ENV CCACHE_COMPRESSLEVEL=0
+ENV CCACHE_MAX_SIZE=5G
+ENV CCACHE_NOHASHDIR=true
```
This seems unnecessary unless we actually have files that're duplicated in the source tree?

jackluo923 replied:
I am confused on your comment. Can you explain how your concern is related to these 4 lines of code?

kirkrodrigues replied:
The comment is attached to line 60; GitHub just shows the previous 3 lines for context.

Why do we want to set CCACHE_NOHASHDIR to true?

jackluo923 replied:
`CCACHE_NOHASHDIR=true` is needed for cache sharing between the warmup build and CI builds. The warmup build runs in `/tmp/velox-src/` while CI builds run in `/__w/velox/velox/`. Without this setting, ccache would treat identical files as different cache entries due to the different working directories, and the warmup cache wouldn't be reused. I've added a comment to clarify this.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+    ccache -vs && \
+    ccache -z
+
+# Remove warmup source files (ccache in /var/cache/ccache is preserved)
```
```suggestion
# Build velox once to warm up ccache
# NOTE:
# - We set `CCACHE_BASEDIR` so cache keys use relative paths.
# - We clear the stats after warmup so that CI builds only show their own cache hits.
COPY . /tmp/velox-src/
WORKDIR /tmp/velox-src
RUN source /opt/velox-venv/bin/activate \
    && CCACHE_BASEDIR=/tmp/velox-src make release \
    && echo "CCache statistics after warmup build:" \
    && ccache --verbose --show-stats \
    && ccache --zero-stats
```

Note that when we break lines, we try to put operators at the start of the line rather than the end of the line.

jackluo923 replied:
The venv is already activated via ENV variables (VIRTUAL_ENV and PATH) set earlier in the Dockerfile, so explicit `source activate` is unnecessary. The ENV-based approach ensures the venv is active for all subsequent RUN commands and any containers started from the image, without needing to remember to source in each RUN.

kirkrodrigues reviewed (commented):
Partial review. @20001020ycx take a look at my comments to see what you could catch in future reviews.

jackluo923 committed:
Update scripts/docker/yscope-velox-builder.dockerfile

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Indent by 4 spaces rather than 2

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Optimized the language in comment

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Optimize the language in comment

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Apply suggestion from @kirkrodrigues

Removed redundant comments

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Apply suggestion from @kirkrodrigues

Convert from 2->4 spaces for indentation.

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
fix: Remove redundant velox source code copying and setting work directory

jackluo923 committed:
refactor(docker): remove redundant venv activation and document ENV approach

Remove the explicit `source activate` in the warmup build RUN command
since the venv is already activated via ENV variables.

Add documentation explaining why we use ENV to set PATH and VIRTUAL_ENV
rather than sourcing the activate script: ENV values persist across all
RUN commands in the Dockerfile and in containers started from the image,
whereas `source activate` only affects a single RUN instruction.

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,77 @@
+# - We set `CCACHE_BASEDIR` so cache keys use relative paths.
+# - We clear the stats after warmup so that CI builds only show their own cache hits.
+COPY . /tmp/velox-src/
+WORKDIR /tmp/velox-src
```
resolved.

jackluo923 committed:
ci: use conditional runner selection for self-hosted vs GitHub-hosted

Use dynamic runner selection based on repository owner:
- y-scope org: use self-hosted runners with appropriate labels
- Other forks: fall back to ubuntu-latest GitHub-hosted runners

This allows forks to run CI without requiring self-hosted runner
infrastructure.

jackluo923 commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,72 @@ on:
+  build-and-test:
+    name: Build and Test (${{ matrix.test-group }})
+    needs: builder-image
+    runs-on: ${{ github.repository_owner == 'y-scope' && fromJSON('["self-hosted", "x64", "ubuntu-noble"]') || 'ubuntu-latest' }}
```
This is fine.

jackluo923 committed:
style(ci): standardize 4-space indentation in runner selection

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+RUN wget --progress=dot:giga https://github.com/Kitware/CMake/releases/download/v3.28.3/cmake-3.28.3-linux-x86_64.sh \
+   && chmod +x cmake-3.28.3-linux-x86_64.sh \
+   && ./cmake-3.28.3-linux-x86_64.sh --skip-license --prefix=/usr/local \
+   && rm cmake-3.28.3-linux-x86_64.sh
```
done

jackluo923 committed:
style(ci): fix indentation per precommit

jackluo923 commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          context: .
+          file: scripts/docker/yscope-velox-builder.dockerfile
+          push: true
+          tags: ${{ steps.image-tag.outputs.tag }}
```
resolved.

kirkrodrigues commented on .github/workflows/linux-build.yml:
```diff
@@ -45,10 +43,16 @@ on:
 
 permissions:
   contents: read
+  # packages: write needed to publish Docker images to GitHub Container Registry (ghcr.io)
```
```suggestion

  # Necessary to publish Docker images to GitHub's container registry (ghcr.io)
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build.yml:
```diff
@@ -45,10 +43,16 @@ on:
+  # Cancel duplicate runs for the same branch. Uses github.head_ref for pull_request (source branch
+  # name, e.g., "ci") and falls back to github.ref_name for push (branch name without refs/heads/).
+  # This ensures both events resolve to the same group (e.g., "Linux Build using GCC - ci").
+  # See: https://github.com/orgs/community/discussions/30794
```
The comments on that discussion say:

> This does not work and is an empty value on PR merge.

jackluo923 replied:
You're right - the previous comment was misleading. `github.head_ref` is empty on PR merge (push event), so the fallback `github.ref_name` would give the target branch ("main"), not the source branch. The comment incorrectly claimed both events resolve to the same group.

I've removed the misleading comment and updated to align with upstream (facebookincubator/velox):
```yaml
group: ${{ github.workflow }}-${{ github.repository }}-${{ github.head_ref || github.sha }}
```

**Behavior change:**
| Scenario | Before (`ref_name` fallback) | After (`sha` fallback) |
|----------|------------------------------|------------------------|
| Cancel redundant PR runs | ✓ Works | ✓ Works |
| Test every merge to main | ✗ May skip some | ✓ All run |
| Save CI resources | ✓ Cancels duplicate main runs | ✗ All runs execute |

**Example of skipped CI (before):** If 3 PRs merge to main in quick succession, all 3 push events share the same concurrency group ("workflow - main"). With `cancel-in-progress: true`, the first two CI runs get cancelled, and only the third one completes - potentially missing bugs introduced by the first two merges.

The new approach ensures every merge to main is tested (safer), at the cost of using more CI resources when multiple PRs merge quickly.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
+  # velox_exec_test takes 15+ minutes (significantly longer than all other tests combined),
+  # so we run it separately to allow other tests to complete quickly in parallel.
+  #
+  # To separate additional slow tests in the future:
```
```suggestion
  # Build and run tests after `builder-image` completes.
  #
  # NOTE: `velox_exec_test` takes 15+ minutes (significantly longer than all other tests combined),
  # so we run it separately to allow other tests to complete quickly in parallel.
  #
  # To separate additional slow tests in the future:
  #
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
+  # 1. Add a new matrix entry with ctest-args: '-R "test_name"'
+  # 2. Update the "other" entry to exclude it: '-E "velox_exec_test|test_name"'
+  build-and-test:
+    name: Build and Test (${{ matrix.test-group }})
```
```suggestion
    name: Build and test (${{ matrix.test-group }})
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
+            ctest-args: -R velox_exec_test
+          - test-group: other
+            ctest-args: -E velox_exec_test
+            # Increase file descriptor limit for velox_table_evolution_fuzzer_test
```
```suggestion

            # Increase file descriptor limit for `velox_table_evolution_fuzzer_test`
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
+            # Increase file descriptor limit for velox_table_evolution_fuzzer_test
+            ulimit-nofile: 8192
+    container:
+      # Image tag is dynamically generated based on dependency hash (intentionally unpinned)
```
Feels unnecessary since if the image is a variable, then it implies the tag is dynamic.

```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
-      run:
-        shell: bash
-        working-directory: velox
+      # Set CCACHE_BASEDIR to checkout path so cache keys use relative paths
```
```suggestion

      # Set `CCACHE_BASEDIR` so cache keys use relative paths
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
-          Boost_SOURCE: SYSTEM
-          ICU_SOURCE: SYSTEM
-          MAKEFLAGS: MAX_HIGH_MEM_JOBS=4 MAX_LINK_JOBS=3
+          # Tune parallelism based on runner type: self-hosted (16-core) vs GitHub-hosted (4-core)
```
```suggestion
          # Tune parallelism based on runner type: Self-hosted (16-core) vs GitHub-hosted (4-core)
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
         run: |
-          if [[ "${USE_CLANG}" = "true" ]]; then export CC=/usr/bin/clang-15; export CXX=/usr/bin/clang++-15; fi
+          # Use Clang compiler if requested (e.g., scheduled.yml uses Clang for compiler diversity)
+          # Otherwise defaults to GCC (default CC/CXX in the container)
```
I feel like this comments are unnecessary since it's really up to the caller why they want to set `use-clang` to `true`.

```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
-      - name: CCache after
-        run: |
-          ccache -vs
+      - name: Show CCache Stats After Build
```
```suggestion
      - name: Show CCache stats after build
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
+            ulimit -n ${{ matrix.ulimit-nofile }}
+          fi
+          cd _build/release && ctest -j $(getconf _NPROCESSORS_ONLN) --output-on-failure --no-tests=error \
+            ${{ matrix.ctest-args }}
```
```suggestion
          cd _build/release \
            && ctest \
              -j $(getconf _NPROCESSORS_ONLN) \
              --output-on-failure \
              --no-tests=error \
              ${{ matrix.ctest-args }}
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+  # Called by other workflows (e.g., linux-build-base.yml) to get the builder image
+  workflow_call:
+    outputs:
+      image-tag:
```
How about `image-ref`? When we say `tag`, people usually think of just the part after the colon (`:`).

jackluo923 replied:
Good point. I've renamed `image-tag` to `image-ref` throughout the codebase. The original naming was influenced by docker/build-push-action's `tags` parameter, which actually expects a full image reference despite its name. Rather than propagating this confusing terminology, I've added a comment at the `tags` parameter to clarify that it expects a full reference, not just the tag portion.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+  workflow_call:
+    outputs:
+      image-tag:
+        description: Fully qualified Docker image tag (ghcr.io/owner/repo/image:hash)
```
```suggestion
        description: Fully qualified Docker image reference (e.g., ghcr.io/owner/repo/image:hash)
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+        value: ${{ jobs.builder-image.outputs.image-tag }}
+  # Allow manual trigger via GitHub Actions UI for rebuilding the image on-demand
+  workflow_dispatch: {}
+  # Automatically rebuild image when dependency installation scripts change
```
Comments feel redundant with code.

```suggestion
  workflow_dispatch: {}

```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          || 'ubuntu-latest'
+      }}
+    outputs:
+      # This output is used by calling workflows to get the image tag
```
```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+        with:
+          persist-credentials: false
+
+      - name: Calculate dependency hash
```
```suggestion
      - name: Hash the dependency scripts
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+        id: deps-hash
+        run: |
+          # Use hashFiles for deterministic hashing of dependency files
+          # Only hash scripts/** to avoid rebuilding when workflow logic changes
```
```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          # Use hashFiles for deterministic hashing of dependency files
+          # Only hash scripts/** to avoid rebuilding when workflow logic changes
+          HASH="${{ hashFiles('scripts/**') }}"
+          # Take first 12 characters for shorter tag (e.g., "a1b2c3d4e5f6")
```
Why truncate?

jackluo923 replied:
The full hash is 64 characters (SHA256), which makes tags harder to read in logs and the GitHub UI. I've changed to 7 characters, which is the industry standard for both Git short hashes and Docker container tags using commit hashes.

References:
- [GitHub Actions: Building Docker Images](https://shipyard.build/blog/gha-recipes-build-and-push-container-registry/) - uses 7-character hashes
- [Docker Best Practices: Using Tags and Labels](https://www.docker.com/blog/docker-best-practices-using-tags-and-labels-to-manage-docker-image-sprawl/)

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          # Use hashFiles for deterministic hashing of dependency files
+          # Only hash scripts/** to avoid rebuilding when workflow logic changes
+          HASH="${{ hashFiles('scripts/**') }}"
+          # Take first 12 characters for shorter tag (e.g., "a1b2c3d4e5f6")
```
```suggestion

          # Truncate to 12 characters for a shorter tag
```

jackluo923 replied:
Changed to 7 characters (Git's default short hash length).

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          # Take first 12 characters for shorter tag (e.g., "a1b2c3d4e5f6")
+          SHORT_HASH="${HASH:0:12}"
+          echo "hash=${SHORT_HASH}" >> "$GITHUB_OUTPUT"
+          echo "Dependency hash: ${SHORT_HASH}"
```
The echo above should be sufficient in GitHub's output, no?

```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+        run: |
+          # GHCR requires lowercase repository names (owner/repo)
+          # Example: "y-scope/velox" -> "y-scope/velox"
+          REPO_LOWER=$(echo "${{ github.repository }}" | tr '[:upper:]' '[:lower:]')
```
```suggestion
          # Docker doesn't support repository names with uppercase characters, so we convert the
          # name to lowercase here.
          REPO_LOWER=$(echo '${{github.repository}}' | tr '[:upper:]' '[:lower:]')"

```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          # Example: "y-scope/velox" -> "y-scope/velox"
+          REPO_LOWER=$(echo "${{ github.repository }}" | tr '[:upper:]' '[:lower:]')
+          # Final tag format: ghcr.io/y-scope/velox/y-scope-velox-builder:a1b2c3d4e5f6
+          TAG="ghcr.io/${REPO_LOWER}/y-scope-velox-builder:${DEPS_HASH}"
```
How about `IMAGE_REF`?

jackluo923 replied:
Applied - renamed to IMAGE_REF.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          # GHCR requires lowercase repository names (owner/repo)
+          # Example: "y-scope/velox" -> "y-scope/velox"
+          REPO_LOWER=$(echo "${{ github.repository }}" | tr '[:upper:]' '[:lower:]')
+          # Final tag format: ghcr.io/y-scope/velox/y-scope-velox-builder:a1b2c3d4e5f6
```
```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+          # Final tag format: ghcr.io/y-scope/velox/y-scope-velox-builder:a1b2c3d4e5f6
+          TAG="ghcr.io/${REPO_LOWER}/y-scope-velox-builder:${DEPS_HASH}"
+          echo "tag=${TAG}" >> "$GITHUB_OUTPUT"
+          echo "Image tag: ${TAG}"
```
```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+        run: |
+          # Use docker manifest inspect to check remote registry without pulling
+          # This is more efficient than docker pull for existence checks
+          # If image with this hash already exists, we can skip the ~60min rebuild
```
```suggestion
          # NOTE: For existence checks, this is more efficient than `docker pull`.
```

jackluo923 replied:
Applied.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+
+      # Only run the following steps if image doesn't exist (conditional on exists == 'false')
+      # Docker Buildx is required for docker/build-push-action to build and push images
+      # It provides advanced build features like multi-stage builds and efficient layer caching
```
```suggestion
```

jackluo923 replied:
Removed.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+      # Only run the following steps if image doesn't exist (conditional on exists == 'false')
+      # Docker Buildx is required for docker/build-push-action to build and push images
+      # It provides advanced build features like multi-stage builds and efficient layer caching
+      - name: Set up Docker Buildx
```
```suggestion
      - name: Extract GitHub metadata
        if: steps.check-image.outputs.exists == 'false'
        uses: docker/metadata-action@v5
        with:
          images: ${{steps.image-tag.outputs.tag}}

      - name: Set up Docker Buildx
```

jackluo923 replied:
Not applied - we use a simpler approach without docker/metadata-action since we only need a single tag based on the dependency hash.

kirkrodrigues commented on .github/zizmor.yml:
```diff
@@ -15,3 +15,15 @@ rules:
   use-trusted-publishing:
     ignore:
       - build_pyvelox.yml
+  # linux-build-base.yml uses ghcr.io/y-scope/docker-github-actions-runner:ubuntu-jammy
```
```suggestion

  # linux-build-base.yml uses ghcr.io/y-scope/docker-github-actions-runner:ubuntu-jammy
```

jackluo923 replied:
Not applied - the zizmor.yml file is for security scanning configuration, not for documenting which images are used where.

kirkrodrigues commented on .github/zizmor.yml:
```diff
@@ -15,3 +15,15 @@ rules:
+      - linux-build-base.yml
+  # yscope-velox-builder-img.yml uses actions/checkout and other first-party GitHub actions
+  # without SHA pinning. For official GitHub-maintained actions, using version tags (e.g., @v4)
+  # allows automatic security patches while maintaining reasonable trust in GitHub's own repos.
```
We can't really trust GitHub and Docker. All companies make mistakes. Even GitHub recommends pinning hashes. Our policy in other repos is to pin to hashes for all actions.

jackluo923 replied:
Understood. I'll keep the current approach for now, but can update to pin hashes if you prefer consistency with other y-scope repos.

jackluo923 replied:
removed

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# limitations under the License.
+
+# Builder image with dependencies and pre-populated ccache for faster CI builds
+FROM ghcr.io/y-scope/docker-github-actions-runner:ubuntu-jammy
```
```suggestion
FROM ghcr.io/myoung34/docker-github-actions-runner:ubuntu-jammy
```

jackluo923 replied:
Applied.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# First, copy the dependency installation scripts to optimize Docker layer caching
+COPY scripts /tmp/velox-deps/
```
For layer caching, wouldn't it make sense to do this *after* setting the timezone and installing CMake---which should rarely change---compared to the scripts---which may change more frequently?

jackluo923 replied:
Applied - moved COPY scripts after timezone and CMake installation for better layer caching.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# PATH and VIRTUAL_ENV via ENV, these values are baked into the Docker image and persist across:
+# 1. All subsequent RUN commands in the Dockerfile
+# 2. Any containers started from the final image (e.g., CI workflows can directly run commands
+#    like `make release` or `python` without needing to activate the venv first)
```
```suggestion
# Activate the virtual environment.
#
# NOTE: We set `ENV` variables directly rather than using `source /opt/velox-venv/bin/activate` in
# a `RUN` command since the latter only persists for that single instruction (each `RUN` starts a
# fresh shell), whereas the former is baked into the Docker image.
```

jackluo923 replied:
Applied.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# CCACHE_MAX_SIZE: 5GB quota (actual usage typically a few hundred MB).
+ENV CCACHE_DIR=/var/cache/ccache
+ENV CCACHE_COMPRESSLEVEL=0
+ENV CCACHE_MAX_SIZE=5G
```
The docs say it's `MAXSIZE` rather than `MAX_SIZE`, no?

```suggestion
ENV CCACHE_MAXSIZE=5G
```

jackluo923 replied:
Applied - fixed to CCACHE_MAXSIZE.

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,76 @@
+# Build velox once to warm up ccache
+# NOTE:
+# - We set `CCACHE_BASEDIR` so cache keys use relative paths.
+# - We clear the stats after warmup so that CI builds only show their own cache hits.
```
```suggestion
# Build velox once to warm up ccache
# NOTE:
# - We set `CCACHE_BASEDIR` so cache keys use relative paths.
# - We clear the stats after warmup so that CI builds only show their own cache hits.
COPY . /tmp/velox-src/
WORKDIR /tmp/velox-src
```

jackluo923 replied:
Applied.

kirkrodrigues replied:
This change hasn't been applied...

kirkrodrigues commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,77 @@ on:
-    name: Ubuntu release with resolve_dependency
-    runs-on: yscope-gh-runner
+  # First job: build or retrieve the Docker builder image
+  # This image contains all dependencies and pre-populated ccache for fast builds
```
```suggestion
```

jackluo923 replied:
Removed.

jackluo923 committed:
Update scripts/docker/yscope-velox-builder.dockerfile

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
refactor(ci): address PR review feedback

- Rename image-tag to image-ref for clarity
- Use 7-char hash (Git standard) instead of 12
- Switch wget to curl with long flags
- Fix CCACHE_MAX_SIZE typo to CCACHE_MAXSIZE
- Reorder Dockerfile layers for better caching
- Align concurrency group with upstream pattern
- Simplify comments and remove redundant ones
- Standardize 4-space indentation throughout

jackluo923 committed:
docs(ci): add comment explaining concurrency model

jackluo923 commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,104 @@
+        with:
+          persist-credentials: false
+
+      - name: Hash the dependency scripts
+        id: deps-hash
+        run: |
+          HASH="${{ hashFiles('scripts/**') }}"
+
+          # Truncate to 7 characters (Git's default short hash length)
+          SHORT_HASH="${HASH:0:7}"
+          echo "hash=${SHORT_HASH}" >> "$GITHUB_OUTPUT"
```
this is fine.

jackluo923 commented on .github/workflows/linux-build-base.yml:
```diff
@@ -24,45 +24,122 @@ on:
+          ccache -z
+
+      - name: Build Velox
+        env:
+          # Tune parallelism based on runner type: self-hosted (16-core) vs GitHub-hosted (4-core)
+          # MAX_HIGH_MEM_JOBS: High-memory compilation tasks (templates, large TUs)
+          # MAX_LINK_JOBS: Parallel linking tasks (memory-intensive)
+          # Self-hosted runners have more cores/RAM, so can handle higher parallelism
+          MAKEFLAGS: >-
+            MAX_HIGH_MEM_JOBS=${{ contains(runner.labels, 'self-hosted') && '16' || '4' }}
+            MAX_LINK_JOBS=${{ contains(runner.labels, 'self-hosted') && '12' || '3' }}
+        run: |
+          # Use Clang compiler if requested (e.g., scheduled.yml uses Clang for compiler diversity)
+          # Otherwise defaults to GCC (default CC/CXX in the container)
+          if [[ "${USE_CLANG}" = "true" ]]; then export CC=/usr/bin/clang-15; export CXX=/usr/bin/clang++-15; fi
+          make release
```
Done - the workflow was refactored to use a single build-and-test job with matrix strategy, eliminating the duplication.

kirkrodrigues commented on .github/workflows/linux-build.yml:
```diff
@@ -46,9 +44,14 @@ on:
+  packages: write
+
+# For PRs: Groups by branch name, so pushing to the same PR cancels previous runs.
+# For pushes to main: Groups by commit SHA, so every merge is tested independently.
```
```suggestion
# For PRs: Group by branch name, so pushing to the same PR cancels previous runs.
# For pushes to main: Group by commit SHA, so every merge is tested independently.
```

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,104 @@
+          context: .
+          file: scripts/docker/yscope-velox-builder.dockerfile
+          push: true
+          # NOTE: `tags` expects a full image reference (e.g., ghcr.io/owner/repo:hash), not just the tag portion
```
```suggestion

          # NOTE: `tags` expects a full image reference (e.g., ghcr.io/owner/repo:hash), not just
          # the part after the colon (':').
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,74 @@
+# in a RUN command since activation in a RUN command only persists for that single instruction
+# (each RUN starts a fresh shell). By setting PATH and VIRTUAL_ENV via ENV, these values are
+# baked into the Docker image and persist across all subsequent RUN commands and any containers
+# started from the final image.
```
```suggestion
# NOTE: We set `ENV` variables directly rather than using `source /opt/velox-venv/bin/activate` in
# a `RUN` command since the latter only persists for that single instruction (each `RUN` starts a
# fresh shell), whereas the former persists across all subsequent `RUN` commands and in containers
# that use the image.
```

kirkrodrigues commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,74 @@
+# Disable compression to trade disk space for speed in CI builds
+ENV CCACHE_COMPRESSLEVEL=0
+ENV CCACHE_MAXSIZE=5G
+# Ignore working directory in cache keys so warmup cache (built in /tmp/velox-src/) is reused
```
```suggestion

# Ignore working directory in cache keys so warmup cache (built in /tmp/velox-src/) is reused
```

jackluo923 committed:
Apply suggestions from code review

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
ci: pin workflow actions to commit SHA with version comments

Pin GitHub actions in linux-build-base.yml and yscope-velox-builder-img.yml
to full commit hashes and add short version comments (e.g. # v4) for
reproducibility and security.

jackluo923 committed:
ci: wire metadata-action labels into builder image build

Add id to Extract GitHub metadata step and pass its labels output to
docker/build-push-action so OCI labels (source, revision, created, etc.)
are embedded in the builder image.

jackluo923 committed:
ci: remove unpinned-uses ignore for yscope-velox-builder-img in zizmor

Workflow actions are now pinned to SHA; drop the zizmor exception and its
comment.

jackluo923 committed:
ci: pin docker/metadata-action to SHA in yscope-velox-builder-img

Satisfy blanket policy requiring actions to be pinned by hash.

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,112 @@
+          echo "ref=${REF}" >> "$GITHUB_OUTPUT"
+
+      - name: Log in to GitHub Container Registry
+        uses: docker/login-action@c94ce9fb468520275223c153574b00df6fe4bcc9 # v3
```
```suggestion
        uses: docker/login-action@c94ce9fb468520275223c153574b00df6fe4bcc9 # v3.7.0
```

Can you do the same for the other ones?

jackluo923 replied:
Used AI to replace all the versions at once.

jackluo923 committed:
Update scripts/docker/yscope-velox-builder.dockerfile

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Update .github/workflows/linux-build.yml

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Update .github/workflows/yscope-velox-builder-img.yml

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
Update scripts/docker/yscope-velox-builder.dockerfile

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>

jackluo923 committed:
ci: use full semver version tags in pinned action comments

kirkrodrigues commented on .github/workflows/yscope-velox-builder-img.yml:
```diff
@@ -0,0 +1,114 @@
+        if: steps.check-image.outputs.exists == 'false'
+        uses: docker/metadata-action@c299e40c65443455700f0fdfc63efafe5b349051 # v5
+        with:
+          images: ${{ steps.image-ref.outputs.ref }}
```
@jackluo923 I think the rabbit is correct about this.

jackluo923 replied:
Comment is below: https://github.com/y-scope/velox/pull/45#issuecomment-3883258323

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,72 @@
+# NOTE: `scripts/setup-ubuntu.sh` installs CMake via pip, but sometimes the pip-installed CMake
+# doesn't show up on the path in container environments (causing, for example, FastFloat library
+# build failures). Using CMake's install script avoids this issue.
+RUN curl --fail --location --show-error --silent --remote-name \
+        https://github.com/Kitware/CMake/releases/download/v3.28.3/cmake-3.28.3-linux-x86_64.sh \
+    && chmod +x cmake-3.28.3-linux-x86_64.sh \
+    && ./cmake-3.28.3-linux-x86_64.sh --skip-license --prefix=/usr/local \
+    && rm cmake-3.28.3-linux-x86_64.sh
```
unnecessary

jackluo923 commented on scripts/docker/yscope-velox-builder.dockerfile:
```diff
@@ -0,0 +1,72 @@
+# limitations under the License.
+
+# Builder image with dependencies and pre-populated ccache for faster CI builds
+FROM ghcr.io/myoung34/docker-github-actions-runner:ubuntu-jammy
```
Do not pin a digest as this image is updated on a daily basis.

jackluo923 commented:
Re: [CodeRabbitAI's comment on `images` in metadata-action](https://github.com/y-scope/velox/pull/45#discussion_r2784347627)

You're correct that passing a tagged ref as `images` is not the intended API usage — fixing this for hygiene.

However, the impact analysis is inaccurate. The generated labels are **not affected** by the `images` input. Looking at the metadata-action source code (`meta.ts:getOCIAnnotationsWithCustoms()`), every label is derived from GitHub context:

| Label | Source |
|-------|--------|
| `org.opencontainers.image.title` | `github.event.repository.name` |
| `org.opencontainers.image.description` | `github.event.repository.description` |
| `org.opencontainers.image.url` | `github.event.repository.html_url` |
| `org.opencontainers.image.source` | `github.event.repository.html_url` |
| `org.opencontainers.image.version` | Derived from tag generation rules + git context |
| `org.opencontainers.image.created` | Current timestamp |
| `org.opencontainers.image.revision` | `github.sha` |
| `org.opencontainers.image.licenses` | Repository license SPDX ID |

**None** of these use the `images` input. The `images` input only affects the `tags` output (which we override explicitly in `build-push-action` anyway). So `org.opencontainers.image.source` is set to the GitHub repo URL (e.g., `https://github.com/y-scope/velox`), not the Docker image reference — it would never have the tag "baked in."

The actual (unused) side-effect is that the `tags` output from metadata-action would be malformed (e.g., `ghcr.io/.../image:hash:ci` — double colon), since the action naively concatenates `imageName + ":" + version`. But since that output isn't consumed, the practical impact is zero.

Fixing anyway for correctness.

jackluo923 committed:
fix(ci): pass base image name to metadata-action's images input

The `images` input of `docker/metadata-action` expects a base image name
without a tag. Split the image-ref step into separate `image` (base name)
and `ref` (tagged) outputs so metadata-action receives the correct input.

kirkrodrigues committed:
Test removing change to zizmor.yml.

kirkrodrigues committed:
Revert "Test removing change to zizmor.yml."

This reverts commit 263cbe8a8c17458a53d2f081e34fcebd5d083824.

kirkrodrigues commented on .github/zizmor.yml:
```diff
@@ -15,3 +15,10 @@ rules:
+  # receive security and maintenance updates automatically
+  unpinned-images:
+    ignore:
+      - linux-build-base.yml
```
```suggestion
  unpinned-images:
    ignore:
      # linux-build-base.yml builds and publishes a container image which we then use immediately;
      # so we need to use the images dynamically generated (i.e., unpinned) tag.
      - linux-build-base.yml
```

kirkrodrigues reviewed (commented):
One more suggestion.

For the PR title _and body,_ how about:

```
ci: Migrate from stateful to stateless runner and apply optimizations to reduce workflow runtime:
```
```
- Build and publish container image for building Velox.
- Switch from stateful runner to stateless runner which uses build-container image.
- Use `ccache` to accelerate workflow runs.
- Parallelize some tests to reduce runtime.
```

jackluo923 committed:
Update .github/zizmor.yml

Co-authored-by: kirkrodrigues <2454684+kirkrodrigues@users.noreply.github.com>
